docs/proposals/daemon-controller.md

```diff
@@ -0,0 +1,151 @@
+  - Modifiers
+    - delete
+    - stop: first we turn down the Daemon Controller foo, and then we turn down all pods matching the query “controller: foo”
+    - label
+    - update
+    - extension: rolling-update
+    - Daemon controllers will have labels, so you could, for example, list all daemon controllers with a certain label (the same way you would for a Replication Controller).
+  - In general, for all the supported features like get, describe, update, etc, the Daemon Controller will work in a similar way to the Replication Controller. However, note that the Daemon Controller and the Replication Controller are different constructs.
+
+### Health checks
+  - Ordinary health checks specified in the pod template will of course work to keep pods created by a Daemon Controller running.
+
+### Cluster Mutations
+  - When a new node is added to the cluster the daemon controller should start the daemon on the node (if the node’s labels match the user-specified selectors). This is a big advantage of the Daemon Controller compared to alternative ways of launching daemons and configuring clusters.
+  - Suppose the user launches a daemon controller that runs a logging daemon on all nodes labeled “tolog”. If the user then adds the “tolog” label to a node (that did not initially have the “tolog” label), the logging daemon should be launched on the node. Additionally, if a user removes the “tolog” label from a node, the logging daemon on that node should be killed.
+
+## Alternatives Considered
+
+An alternative way to launch daemons is to avoid going through the API server, and instead provide ways to package the daemon into the node. For example, users could:
+
+1. Include the daemon in the machine image
+2. Use config files to launch daemons
+3. Use static pod manifests to launch daemon pods when the node initializes
+
+These alternatives don’t work as well because the daemons won’t be well integrated into Kubernetes. In particular,
+
+1. In alternatives (1) and (2), health checking for the daemons would need to be re-implemented, or would not exist at all (because the daemons are not run inside pods). In the current proposal, the Kubelet will health-check daemon pods and restart them if necessary.
+2. In alternatives (1) and (2), binding services to a group of daemons is difficult (which is needed in use cases such as the sharded data store use case described above), because the daemons are not run inside pods
+3. A big disadvantage of these methods is that adding new daemons in existing nodes is difficult (for example, if a cluster manager wants to add a logging daemon after a cluster has been deployed).
+4. The above alternatives are less user-friendly. Users need to learn two ways of launching pods: using the API when launching pods associated with Replication Controllers, and using manifests when launching daemons. So in the alternatives, deployment is more difficult.
+5. It’s difficult to upgrade binaries launched in any of those three ways.
```

Comment:
Because pods have to explicitly declare their dependencies (e.g. use hostDir to explicitly mount part of the node FS, use hostPort, etc), it is somewhat easier to reason about interactions between daemons, and prevents new dependencies from creeping in.  This makes independently updating daemons safer.